gap/bliss.gi

```diff
@@ -12,141 +13,247 @@
+function(gr1, gr2)
+  local act;
+
+  if gr1 = gr2 then
+    return true;
+  fi;
 
   # check some invariants
-  if DigraphNrVertices(g1) <> DigraphNrVertices(g2)
-      or DigraphNrEdges(g1) <> DigraphNrEdges(g2)
-      or IsMultiDigraph(g1) <> IsMultiDigraph(g2) then
+  if DigraphNrVertices(gr1) <> DigraphNrVertices(gr2)
+      or DigraphNrEdges(gr1) <> DigraphNrEdges(gr2)
+      or IsMultiDigraph(gr1) <> IsMultiDigraph(gr2) then
     return false;
   fi; #JDM more!
 
-  if IsMultiDigraph(g1) then
-    return OnMultiDigraphs(g1, DigraphCanonicalLabelling(g1))
-      = OnMultiDigraphs(g2, DigraphCanonicalLabelling(g2));
+  if IsMultiDigraph(gr1) then
+    act := OnMultiDigraphs;
+  else
+    act := OnDigraphs;
   fi;
-  return OnDigraphs(g1, DigraphCanonicalLabelling(g1))
-    = OnDigraphs(g2, DigraphCanonicalLabelling(g2));
+
+  return act(gr1, DigraphCanonicalLabelling(gr1))
+    = act(gr2, DigraphCanonicalLabelling(gr2));
+end);
+
+#
+
+InstallMethod(IsIsomorphicDigraph, "for digraphs and homogeneous lists",
+[IsDigraph, IsDigraph, IsHomogeneousList, IsHomogeneousList],
+function(gr1, gr2, c1, c2)
+  local m, colour1, n, colour2, class_sizes, i;
+
+  if IsMultiDigraph(gr1) or IsMultiDigraph(gr2) then
+    ErrorNoReturn("Digraphs: IsIsomorphicDigraph: usage,\n",
+                  "isomorphism testing is not supported for coloured ",
+                  "digraphs with multiple\nedges,");
+  fi;
+  m := DigraphNrVertices(gr1);
+  colour1 := DIGRAPHS_ValidateVertexPartition(m, c1);
+  if colour1 = fail then
+    ErrorNoReturn("Digraphs: IsIsomorphicDigraph: usage,\n",
+                  "the argument <c1> does not define a valid colouring of the ",
+                  "vertices of <gr1>.\nEither <c1> should be a list of ",
+                  "length <m>, consisting of integers in the\nrange [0 .. <m>",
+                  "], or <c1> should be a list of non-empty disjoint ",
+                  "lists whose\nunion is the vertices of <gr1>,");
+  fi;
+  n := DigraphNrVertices(gr2);
+  colour2 := DIGRAPHS_ValidateVertexPartition(n, c2);
+  if colour2 = fail then
+    ErrorNoReturn("Digraphs: IsIsomorphicDigraph: usage,\n",
+                  "the argument <c2> does not define a valid colouring of the ",
+                  "vertices of <gr2>.\nEither <c2> should be a list of ",
+                  "length <n>, consisting of integers in the\nrange [0 .. <n>",
+                  "], or <c2> should be a list of non-empty disjoint ",
+                  "lists whose\nunion is the vertices of <gr2>,");
+  fi;
+
+  # check some invariants
+  if m <> n or DigraphNrEdges(gr1) <> DigraphNrEdges(gr2) then
+    return false;
+  fi;
+
+  class_sizes := ListWithIdenticalEntries(n, 0);
+  for i in DigraphVertices(gr1) do
+    class_sizes[colour1[i]] := class_sizes[colour1[i]] + 1;
+    class_sizes[colour2[i]] := class_sizes[colour2[i]] - 1;
+  od;
+  if not ForAll(class_sizes, x -> x = 0) then
+    return false;
+  elif gr1 = gr2 and colour1 = colour2 then
```

Comment:
Do we have any evidence that equality checking is faster than checking isomorphism? I don't really have any preconceptions, except that finding a canonical labelling is usually very fast, so I'm just checking.